src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -159,61 +225,76 @@ absl::optional<int64_t> GcsTaskManager::GcsTaskManagerStorage::GetTaskStatusUpda
+    const TaskID &task_id, int64_t failed_ts, const TaskID &ancestor_task_id) {
+  RAY_LOG(DEBUG) << "Marking task " << task_id << " as failed due to ancestor task "
+                 << ancestor_task_id << " failure.";
   auto latest_task_attempt = GetLatestTaskAttempt(task_id);
   if (!latest_task_attempt.has_value()) {
     return;
   }
-  MarkTaskAttemptFailed(*latest_task_attempt, failed_ts);
+  rpc::RayErrorInfo error_info;
+  error_info.set_error_type(rpc::ErrorType::WORKER_DIED);
+  std::stringstream error_message;
+  error_message << "Ancestor task " << ancestor_task_id
+                << " failed due to unexpected worker death.";
+  error_info.set_error_type(rpc::ErrorType::WORKER_DIED);
+  error_info.set_error_message(error_message.str());
+
+  MarkTaskAttemptFailed(*latest_task_attempt, failed_ts, error_info);
 }
 
 void GcsTaskManager::GcsTaskManagerStorage::MarkTaskTreeFailedIfNeeded(
-    const TaskID &task_id, const TaskID &parent_task_id) {
-  if (!parent_task_id.IsNil()) {
-    // If parent has failed, mark itself as failed
-    auto parent_failed_ts =
-        GetTaskStatusUpdateTime(parent_task_id, rpc::TaskStatus::FAILED);
-    if (parent_failed_ts.has_value()) {
-      // Mark current task as failed.
-      MarkTaskFailed(task_id, *parent_failed_ts);
-    }
+    const TaskID &task_id) {
+  const auto &task_event = GetLatestTaskEvent(task_id);
+  if (!task_event.has_value() || !task_event->has_task_info()) {
+    return;
   }
-
-  // BFS traverse the task tree to mark all non terminal children as failure
-  std::vector<TaskID> failed_tasks;
-  auto task_failed_ts = GetTaskStatusUpdateTime(task_id, rpc::TaskStatus::FAILED);
-  if (task_failed_ts.has_value()) {
-    failed_tasks.push_back(task_id);
+  auto parent_task_id = TaskID::FromBinary(task_event->task_info().parent_task_id());
+  // BFS traverse the task tree to mark all non terminal children as failure if needed.
+  std::vector<TaskID> parent_tasks_to_check;
+  parent_tasks_to_check.push_back(task_id);
+  if (!parent_task_id.IsNil()) {
+    parent_tasks_to_check.push_back(parent_task_id);
+  } else {
+    parent_tasks_to_check.push_back(task_id);
   }
 
-  for (size_t i = 0; i < failed_tasks.size(); ++i) {
-    auto failed_task_id = failed_tasks[i];
-    auto children_tasks_itr = parent_to_children_task_index_.find(failed_task_id);
+  for (size_t i = 0; i < parent_tasks_to_check.size(); ++i) {
+    auto parent_task_id = parent_tasks_to_check[i];
+    auto children_tasks_itr = parent_to_children_task_index_.find(parent_task_id);
     if (children_tasks_itr == parent_to_children_task_index_.end()) {
       continue;
     }
     for (const auto &child_task_id : children_tasks_itr->second) {
       // Mark any non-terminated child as failed with parent's failure timestamp.
-      if (!IsTaskTerminated(child_task_id)) {
-        MarkTaskFailed(child_task_id, task_failed_ts.value());
-        failed_tasks.push_back(child_task_id);
+      int64_t task_failed_ts = 0;
```

Comment:
Is it possible to set task_failed_ts outside `ShouldMarkChildrenFailed` method?

Reply:
yes, i am not a fan of this also. maybe return the failed ts, and -1 as a falsy value?